File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {

	reader := bufio.NewReader(os.Stdin)
	var n, m int
	var w float64

	fmt.Fscanf(reader, "%d %d\n", &n, &m)
	var graph = make([][]float64, n)
	var res = make([][]float64, n)
	for i := 0; i < n; i++ {
		graph[i] = make([]float64, m)
		res[i] = make([]float64, m)
		var s string
		fmt.Fscanf(reader, "%s\n", &s)
		sArr := strings.Split(s, "")
		for j := 0; j < len(sArr); j++ {
			a, _ := strconv.Atoi(sArr[j])
			graph[i][j] = float64(a)
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if i == 0 || j == 0 {
				res[i][j] = graph[i][j]
			} else if graph[i][j] == 0 {
				res[i][j] = 0
			} else {
				res[i][j] = math.Min(math.Min(res[i-1][j-1], res[i-1][j]), res[i][j-1]) + 1
			}
			w = math.Max(res[i][j], w)
		}
	}
	fmt.Printf("%0.f\n", w*w)
}

File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

n, m = map(int, input().split())
graph = [list(map(int, list(input().strip()))) for _ in range(n)]
res = [[0]*m for _ in range(n)]
w = 0
for i in range(n):
    for j in range(m):
        if i == 0 or j == 0:
            res[i][j] = graph[i][j]
        elif graph[i][j] == 0:
            res[i][j] = 0
        else:
            res[i][j] = min(res[i - 1][j - 1], res[i - 1][j], res[i][j - 1]) + 1
        w = max(res[i][j], w)
print(w*w)